java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -96,93 +142,211 @@ public void onTimer(
             KeyedProcessFunction<Row, Row, Row>.OnTimerContext ctx,
             Collector<Row> out)
             throws Exception {
-        aggregationFieldsDescriptor.getAggFieldDescriptors().forEach(d -> d.aggFunc.reset());
 
         boolean hasRow = false;
 
-        for (long rowTime : state.keyToTimestamps.get(ctx.getCurrentKey())) {
-            if (rowTime > timestamp) {
-                break;
+        final List<Long> timestampList = state.getTimestampList();
+        final int timestampListSize = timestampList.size();
+
+        Row outputRow = Row.withNames();
+        for (int i = 0; i < keyFieldNames.length; i++) {
+            outputRow.setField(keyFieldNames[i], ctx.getCurrentKey().getField(i));
+        }
+
+        Row rowToAdd = state.timestampToRow.get(timestamp);
+        final Row accumulators = state.getAccumulators();
+        final Map<String, Integer> leftIdxMap = state.getLeftTimestampIdx();
+
+        for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
+                aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+            Object accumulator = accumulators.getField(descriptor.outFieldName);
+            if (accumulator == null) {
+                accumulator = descriptor.aggFunc.createAccumulator();
+                accumulators.setField(descriptor.outFieldName, accumulator);
+            }
+            if (rowToAdd != null) {
+                descriptor.aggFunc.add(
+                        accumulator, rowToAdd.getField(descriptor.inFieldName), timestamp);
             }
-            Row curRow = state.timestampToRow.get(rowTime);
 
-            // TODO: Optimize by supporting retract value from aggregation function.
-            for (AggregationFieldsDescriptor.AggregationFieldDescriptor descriptor :
-                    aggregationFieldsDescriptor.getAggFieldDescriptors()) {
-                if (timestamp - descriptor.windowSizeMs >= rowTime) {
-                    continue;
+            // Advance left idx and retract values
```

Comment:
Advance left idx and retract values

-> 

Advance left idx and retract rows whose rowTime is out of the time window `(timetamp - descriptor.windowSizeMs, timestamp]`.